integration-tests/tests/src/tests/dictionary.ts

```diff
@@ -0,0 +1,315 @@
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+import { expect } from "chai";
+import Realm from "realm";
+
+import { openRealmBefore } from "../hooks";
+
+type Item<ValueType = Realm.Mixed> = {
+  dict: Realm.Dictionary<ValueType>;
+};
+
+type DictValues = { [key: string]: unknown };
+
+type ValueGenerator = (realm: Realm) => DictValues;
+
+type TypedDictionarySuite = {
+  extraSchema?: Realm.ObjectSchema[];
+  type: string;
+  goodValues: DictValues | ValueGenerator;
+  badValues: DictValues;
+  expectedError: string;
+};
+
+describe("Dictionary", () => {
+  describe("with unconstrained (mixed) values", () => {
+    openRealmBefore({
+      schema: [
+        {
+          name: "Item",
+          properties: { dict: "{}" },
+        },
+        {
+          name: "Person",
+          properties: { name: "string" },
+        },
+      ],
+    });
+
+    it("can be used as a property type in a schema", function (this: RealmContext) {
+      expect(this.realm.isClosed).equals(false);
+      const dictSchemaProperty = this.realm.schema[0].properties.dict as Realm.ObjectSchemaProperty;
+      expect(typeof dictSchemaProperty).equals("object");
+      expect(dictSchemaProperty.type).equals("dictionary");
+      expect(dictSchemaProperty.objectType).equals("mixed");
+    });
+
+    it("is instanceof Dictionary", function (this: RealmContext) {
+      this.realm.write(() => {
+        const item = this.realm.create<Item>("Item", {});
+        expect(item.dict instanceof Realm.Dictionary);
+      });
+    });
+
+    const methodNames = [
+      // "get",
+      "set",
+      "remove",
+      "addListener",
+      "removeListener",
+      "removeAllListeners",
+    ];
+    for (const name of methodNames) {
+      it(`expose a method named '${name}'`, function (this: RealmContext) {
+        this.realm.write(() => {
+          const item = this.realm.create<Item>("Item", {});
+          expect(typeof item.dict[name]).equals("function");
+        });
+      });
+    }
+
+    it("can store string values using string keys", function (this: RealmContext) {
+      this.realm.write(() => {
+        const item = this.realm.create<Item>("Item", {});
+        item.dict.key1 = "hello";
+        expect(item.dict.key1).equals("hello");
+      });
+    });
+
+    it("can store number values using string keys", function (this: RealmContext) {
+      this.realm.write(() => {
+        // Creation
+        const item = this.realm.create<Item<number>>("Item", {
+          dict: { key1: 0 },
+        });
+        expect(item.dict).deep.equals({ key1: 0 });
+        // Assignment
+        item.dict.key1 = 1234;
+        item.dict.key2 = Number.MAX_VALUE;
+        item.dict.key3 = Number.MIN_VALUE;
+        expect(item.dict).deep.equals({
+          key1: 1234,
+          key2: Number.MAX_VALUE,
+          key3: Number.MIN_VALUE,
+        });
+        // Increments
+        item.dict.key1++;
```

Comment:
Is this a c++ operation?  If so shouldn't we also test `-- += -=`?  Also what happens with `item.dict.key1 + "one"`?